packages/server-wallet/src/wallet/store.ts

```diff
@@ -1,66 +1,225 @@
     // TODO: Implement this
-    return Promise.resolve();
+    return Promise.resolve(right(undefined));
   },
-  addSignedState: async function(signedState: SignedState): Promise<void> {
-    if (!signedState.signatures?.length) {
-      throw new Error('No signature for signed state');
-    }
+  addSignedState: async function(signedState: SignedState): Promise<Either<StoreError, undefined>> {
+    const sigValidationResult = validateSignatures(signedState);
+    if (isLeft(sigValidationResult)) return sigValidationResult;
 
-    const channelId = calculateChannelId(signedState);
-    let channel = await Channel.query()
-      .where('channelId', channelId)
-      .first();
-
-    if (!channel) {
-      const addresses = signedState.participants.map(p => p.signingAddress);
-      const signingWallet = await SigningWallet.query()
-        .whereIn('address', addresses)
-        .first();
-
-      if (!signingWallet) {
-        logger.error(
-          {
-            knownWallets: await SigningWallet.query().select(),
-            addresses,
-          },
-          'Not in channel'
-        );
-        throw Error('Not in channel');
-      }
-
-      const {address: signingAddress} = signingWallet;
-      const cols: RequiredColumns = {...signedState, vars: [addHash(signedState)], signingAddress};
-
-      channel = Channel.fromJson(cols);
-      await Channel.query().insert(channel);
-    } else {
-      signedState.signatures?.map(sig => channel.addState(signedState, sig));
-      await Channel.query().update(channel);
-    }
+    const singingWalletResult = await getSigningWallet(signedState);
+    if (isLeft(singingWalletResult)) return singingWalletResult;
+
+    const {address: signingAddress} = singingWalletResult.right;
+
+    const cols: RequiredColumns = {...signedState, vars: [addHash(signedState)], signingAddress};
+    const channel = Channel.fromJson(cols);
+
+    let channelVars = channel.vars;
+
+    const addResult = addState(channelVars, signedState);
+    if (isLeft(addResult)) throw addResult.left;
+
+    channelVars = clearOldStates(channelVars, channel.isSupported ? channel.support : undefined);
+
+    const invariantValidationResult = validateInvariants(channelVars, channel.myAddress);
+    if (isLeft(invariantValidationResult)) return invariantValidationResult;
+
+    await Channel.query().upsertGraph(channel);
+
+    return right(undefined);
   },
 };
+
+class StoreError extends Error {
+  readonly type = 'InvariantError';
+  constructor(reason: StoreErrors, public readonly data: any = undefined) {
+    super(reason);
+  }
+}
+
+enum StoreErrors {
+  duplicateTurnNums = 'multiple states with same turn number',
+  notSorted = 'states not sorted',
+  multipleSignedStates = 'Store signed multiple states for a single turn',
+  invalidSignature = 'Invalid signature',
+  notInChannel = 'Not in channel',
+  staleState = 'Stale state',
+}
```

Comment:
Is there some way we can make these things be parametrizable functions by default? i.e., when defining a new exception I can define both the exception type and the arguments that led to it? e.g., `multipleSignedStates(ss1: SignedState, ...sss: SignedState[])`?